extensions/jms/src/main/java/org/mule/extensions/jms/api/connection/JmsConnection.java

```diff
@@ -0,0 +1,313 @@
+
+import static java.lang.String.format;
+import static org.apache.commons.lang.RandomStringUtils.randomAlphanumeric;
+import static org.apache.commons.lang.StringUtils.isBlank;
+import static org.mule.extensions.jms.api.config.AckMode.MANUAL;
+import static org.mule.extensions.jms.api.config.AckMode.TRANSACTED;
+import org.mule.extensions.jms.api.config.AckMode;
+import org.mule.extensions.jms.api.destination.ConsumerType;
+import org.mule.extensions.jms.internal.support.JmsSupport;
+import org.mule.runtime.api.exception.MuleException;
+import org.mule.runtime.api.lifecycle.Disposable;
+import org.mule.runtime.api.lifecycle.Stoppable;
+
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+import javax.jms.Connection;
+import javax.jms.Destination;
+import javax.jms.JMSException;
+import javax.jms.Message;
+import javax.jms.MessageConsumer;
+import javax.jms.MessageProducer;
+import javax.jms.QueueSender;
+import javax.jms.Session;
+import javax.jms.Topic;
+import javax.jms.TopicPublisher;
+import javax.jms.TopicSession;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A Connection for the JmsExtension
+ *
+ * @since 4.0
+ */
+public final class JmsConnection implements Stoppable, Disposable {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(JmsConnection.class);
+
+  private final JmsSupport jmsSupport;
+  private final Connection connection;
+  private final Map<String, Message> pendingAckSessions = new HashMap<>();
+  private final List<MessageConsumer> createdConsumers = new LinkedList<>();
+  private final List<MessageProducer> createdProducers = new LinkedList<>();
+  private final List<Session> createdSessions = new LinkedList<>();
+
+  public JmsConnection(JmsSupport jmsSupport, Connection connection) {
+    this.jmsSupport = jmsSupport;
+    this.connection = connection;
+  }
+
+  public JmsSupport getJmsSupport() {
+    return jmsSupport;
+  }
+
+  public Connection get() {
+    return connection;
+  }
+
+  /**
+   * Creates a new JMS {@link Session} using the current {@link Connection}
+   *
+   * @param ackMode the {@link Session} {@link AckMode}
+   * @param isTopic if {@code true} the {@link Session} created will be a {@link TopicSession}.
+   *                This distinction is made only for {@link JmsSpecification#JMS_1_0_2b}
+   * @return a new {@link Session}
+   * @throws JMSException if an error occurs while creating the {@link Session}
+   */
+  public JmsSession createSession(AckMode ackMode, boolean isTopic) throws JMSException {
+    Session session = jmsSupport.createSession(connection, isTopic, ackMode.equals(TRANSACTED), ackMode.getAckMode());
+    createdSessions.add(session);
+
+    if (ackMode.equals(MANUAL)) {
+      String ackId = randomAlphanumeric(16);
+      pendingAckSessions.put(ackId, null);
+      return new JmsSession(session, ackId);
+    }
+
+    return new JmsSession(session, "");
```

Comment:
What's the meaning of creating a JmsSession with "" as ack ID? If ack ID is optional, make it optional.